### DISCUSSION TITLE
Error handling in root.tsx as soon as loader is in use

### DISCUSSION THREAD

**Discussion by bartylla** [remix-run/remix#9681] [Q&A] [answered]
I started a new project by running `npx create-remix@latest --template remix-run/remix/templates/express`.

In the `root.tsx` I want to provide a few generic information via loader to the layout. So I added a loader function and the useLoaderData hook.

```TS
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useLoaderData,
} from "@remix-run/react";
import "./tailwind.css";
import { LoaderFunctionArgs } from "@remix-run/node";

export const loader = async ({ request }: LoaderFunctionArgs) => {
  return {
    message: "loader",
  };
};

export function Layout({ children }: { children: React.ReactNode }) {
  const { message } = useLoaderData<typeof loader>();

  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}

export default function App() {
  return <Outlet />;
}
```

In general this is working fine as long no errors occur somewhere else. As soon I open an unknown route like localhost:3000/test I get the following error:

```
No routes matched location "/test" 
You cannot `useLoaderData` in an errorElement (routeId: root)
TypeError: Cannot destructure property 'test' of '__vite_ssr_import_3__.useLoaderData(...)' as it is undefined.
```

Without the loader I get a 404 screen, and on the index page I get the value is available like expected.
Would be great if someone could tell me what I'm doing wrong here or how I can catch the error properly.

**Comment by AlemTuzlak** [comment, marked as answer]
The `Layout` element is supposed to be used to render html markup in both the case of an error and a valid page, which means you can't use a `useLoaderData` hook inside of it because in case of an `ErrorBoundary` it doesn't exist. Try moving your useLoaderData to the `App` export. Whatever you return from the `App` will be rendered inside the `Layout` as `children`

**Reply by bartylla** [reply]
Thanks for the explanation. I adjusted the code like below and it is working now.

Since I merged the Layout and App component to fix the problem. What is the benefit to have the two components in the root file by default?

```TS
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useLoaderData,
} from "@remix-run/react";
import "./tailwind.css";
import { LoaderFunctionArgs } from "@remix-run/node";

export const loader = async ({ request }: LoaderFunctionArgs) => {
  return {
    message: "loader",
  };
};

export default function App() {
  const { message } = useLoaderData<typeof loader>();

  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}
```

**Reply by AlemTuzlak** [reply]
If you look at this documentation piece:
https://remix.run/docs/en/main/file-conventions/root#layout-export
You can see this piece here:
```tsx

export default function App() {
  return <Outlet />;
}

export function ErrorBoundary() {
  const error = useRouteError();

  if (isRouteErrorResponse(error)) {
    return (
      <>
        <h1>
          {error.status} {error.statusText}
        </h1>
        <p>{error.data}</p>
      </>
    );
  }

  return (
    <>
      <h1>Error!</h1>
      <p>{error?.message ?? "Unknown error"}</p>
    </>
  );
}

```

If it didn't export the `Layout` you would have to copy/paste:
```tsx
  <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
```

Into both the `App` export and `ErrorBoundary` but this allows you to not do that, BUT with the special case you can't use `useLoaderData` or `useRouteError` inside the layout export itself
